### ISSUE TITLE
Implement Universal Shaping Engine

### ISSUE THREAD

**Issue by devongovett** [devongovett/fontkit#53]
Spec: https://www.microsoft.com/typography/OpenTypeDev/USE/intro.htm
Harfbuzz: https://github.com/behdad/harfbuzz/blob/master/src/hb-ot-shape-complex-use.cc

**Comment by devongovett** [comment]
Hacked on a state machine compiler a bit: https://github.com/devongovett/state-machine. It compiles a regular expression like syntax to a DFA. It's similar to Ragel, which Harfbuzz uses. All of the JS backends for Ragel I found produced output with a pretty large code size. The compiler I wrote produces a minimal state table which can be serialized to JSON, and has a very small runtime (< 100 lines). Not sure what to call it though: all the good names are taken in npm...

This should make it much easier to implement the syllable cluster pattens from the OT spec, rather than by hand. We should be able to use the machine from Harfbuzz (which is in turn very close to the spec) without many modifications.
